### ISSUE TITLE
`--distdir` and `--ghcjs` don’t go well together

### ISSUE THREAD

**Issue by nomeata** [haskell/cabal#5453]
In https://travis-ci.org/nomeata/incredible/builds/407350923 I stumbled over this:
```
cd logic && cabal new-build --distdir=dist-newstyle-ghcjs --ghcjs --dependencies-only --disable-tests
cabal: '/opt/ghcjs/8.2/bin/ghcjs' exited with an error:
ghcjs: can't find a package database at
/home/travis/build/nomeata/incredible/logic/dist-newstyle/packagedb/ghc-8.2.2
```
Why is it not adhering to the `--distdir` that I gave?

This is using `cabal-install-head_2.3+git20180723.1.67bdbab-6~14.04_amd64.deb` … and weirdly it works locally (with a self-built `ghcjs` though)

 @hvr any ideas?

**Comment by nomeata** [comment]
Yes, that was it. so I guess this is related to https://github.com/haskell/cabal/issues/4542

One could argue that if a user specifies a `--distdir`, he wants to achieve some form of isolation, and no environment file should be written. Or maybe it should just be the case that all invocations of the compiler through `cabal` should ignore the environment file (after all, `cabal` knows better what the flags ought to be.)

**Comment by 23Skidoo** [comment]
So GHCJS picks up `.ghc.environment.*` files? IMO it should use `ghcjs.environment.*`.

/cc @luite

**Comment by hvr** [comment]
@23Skidoo yes, unfortunately GHCJS does pick up the wrong env files; there would be two ways to address this (probably both should be taken):

1. Have `ghcjs` set a proper architecture triple; currently it's lying; see `ghcjs --info` below which have it claim to be a native compiler for e.g. Linux (`("target arch","ArchX86_64")` & `("target os","OSLinux")`) and which contains entries which make no sense for ghcjs; this requires hacks in tooling to ignore those for GHCJS;  or also see how cabal new-build names the build dir:

       ./dist-newstyle/build/x86_64-linux/ghcjs-8.4.0.1/dlist-0.8.0.4/build/libHSdlist-0.8.0.4-inplace.js_a

2. or alternatively, have `ghcjs` properly rename occurrences of `ghc` into `ghcjs` (i.e. as you suggest, `.ghc.environment.*` to `ghcjs.environment.*` but also renaming the env-var `GHC_ENVIRONMENT` accordingly)

## `ghcjs --info` output

```hs
 [("Project name","The Glorious Glasgow Haskell Compilation System for JavaScript")
 ,("Global Package DB","/opt/ghcjs/8.4.0.1/lib/ghcjs/package.conf.d")
 ,("Project version","8.4.0.1")
 ,("LibDir","/opt/ghcjs/8.4.0.1/lib/ghcjs")
 ,("Native Too","NO")
 ,("GCC extra via C opts"," -fwrapv -fno-builtin")
 ,("C compiler command","gcc")
 ,("C compiler flags"," -fno-stack-protector")
 ,("C compiler link flags"," -fuse-ld=gold")
 ,("C compiler supports -no-pie","YES")
 ,("Haskell CPP command","gcc")
 ,("Haskell CPP flags","-E -undef -traditional")
 ,("ld command","ld.gold")
 ,("ld flags","")
 ,("ld supports compact unwind","YES")
 ,("ld supports build-id","YES")
 ,("ld supports filelist","NO")
 ,("ld is GNU ld","YES")
 ,("ar command","ar")
 ,("ar flags","q")
 ,("ar supports at file","YES")
 ,("ranlib command","ranlib")
 ,("touch command","touch")
 ,("dllwrap command","/bin/false")
 ,("windres command","/bin/false")
 ,("libtool command","libtool")
 ,("perl command","/usr/bin/perl")
 ,("cross compiling","NO")
 ,("target os","OSLinux")
 ,("target arch","ArchX86_64")
 ,("target word size","8")
 ,("target has GNU nonexec stack","True")
 ,("target has .ident directive","True")
 ,("target has subsections via symbols","False")
 ,("target has RTS linker","YES")
 ,("Unregisterised","NO")
 ,("LLVM llc command","llc")
 ,("LLVM opt command","opt")
 ,("LLVM clang command","clang")
 ,("Project Git commit id","0000000000000000000000000000000000000000")
 ,("Booter version","8.4.1.20180414")
 ,("Stage","2")
 ,("Build platform","x86_64-unknown-linux")
 ,("Host platform","x86_64-unknown-linux")
 ,("Target platform","x86_64-unknown-linux")
 ,("Have interpreter","YES")
 ,("Object splitting supported","YES")
 ,("Have native code generator","YES")
 ,("Support SMP","YES")
 ,("Tables next to code","YES")
 ,("RTS ways","l debug thr thr_debug thr_l thr_p dyn debug_dyn thr_dyn thr_debug_dyn l_dyn thr_l_dyn")
 ,("RTS expects libdw","NO")
 ,("Support dynamic-too","YES")
 ,("Support parallel --make","YES")
 ,("Support reexported-modules","YES")
 ,("Support thinning and renaming package flags","YES")
 ,("Support Backpack","YES")
 ,("Requires unified installed package IDs","YES")
 ,("Uses package keys","YES")
 ,("Uses unit IDs","YES")
 ,("Dynamic by default","NO")
 ,("GHC Dynamic","NO")
 ,("GHC Profiled","NO")
 ,("Leading underscore","NO")
 ,("Debug on","False")
 ]
```

**Comment by hvr** [comment]
@nomeata

> Or maybe it should just be the case that all invocations of the compiler through cabal should ignore the environment file (after all, cabal knows better what the flags ought to be.)

that should actually be the case (at least for the `new-*` commands; I wouldn't be surprised if the legacy old-build ones might have some gap somewhere; but w/ new-* I don't think I ran into any issues; also not that iirc `new-exec` requires ghc env files for its operation).

**Comment by nomeata** [comment]
> that should actually be the case (at least for the new-* commands)

Hmm, the command line that shows this problem at the very top *is* using `new-build`.

**Comment by hvr** [comment]
@nomeata oh, indeed!

this points to something more fundamental: `ghc --info` (which is called early on by Cabal to probe for capabilities) happily ignores any .ghc.environment files; `ghcjs --info` however does *not* (but Cabal expects it to!)

`ghcjs --info` seems to invoke additional codepaths that `ghc --info` does not invoke which trigger checking for package environments... I haven't yet looked into why/where this happens.

And to make things worse, `.ghc.environment.*` files are not supposed to be looked for when the `-hide-all-packages` flag is passed... but `ghcjs` seems to happily look em up and crash & burn ... 

long story short: the fault is cliearly on GHCJS' end; so the only thing we can do is to fix this in GHCJS upstream asap

**Comment by nomeata** [comment]
Great analysis, thanks!
